Replace fast_csr_matrix in core

jakelishman opened the pull request:
_Edited 2020-08-10_

This is now a complete switch-over to the new `CSR` type, from `fast_csr_matrix`.  There are huge changes and knock-on effects all across QuTiP.

Most notably for user code is that arithmetic on `Qobj` is now massively faster for small `Qobj`.  The instantiation cost of `Qobj` is reduced an awful lot, and internal functions which return `Qobj` now supply `dims` and `type` information to the constructor so that instantiation time is down from ~50-100µs to about ~5µs.

Notable removals:
 - `eseries` and `essolve` are completely gone
 - OpenMP is temporarily removed (though this will be added back in to the data layer later)
 - all functions in `correlation` and `qobj` which were marked "deprecated" are now actually removed
 - `vec2mat` and `mat2vec` have been renamed to `vector_to_operator` and `operator_to_vector`
 - all "legacy" files have been completely removed
 - all previous scipy-CSR-like Cython code is completely gone (replaced by the data layer)
 - `Qobj` no longer defines `__array__`, and so will not automatically turn itself into one.  Having this produced very inconsistent behaviour when multiplying by numpy arrays, or otherwise dealing with numpy functions.  If you _need_ `Qobj` to be array, be explicit and call `Qobj.full()`.

Notable changes in `Qobj`:
 - matrix multiplication now contracts to a scalar if doing `bra @ ket`
 - matrix multiplication no longer contracts 1D subspaces, but the new `Qobj.contract()` will do this
 - arithmetic operations are much more strict about enforcing `Qobj.type`-safety
 - `Qobj` understands the `@` operator to mean matrix multiplication, but also maintains the same behaviour of treating `Qobj * Qobj` as matrix multiplication

Package organisation:
 - the solvers are now physically contained in `qutip.solve`, but are still logically available in the `qutip` namespace.  This means that direct imports (like `from qutip.mesolve import mesolve`) will fail (it becomes `from qutip.solve.mesolve import mesolve`), but the "proper" unqualified access `from qutip import mesolve` remains unchanged.  This is for internal code organisation purposes.

The integral type used as the indices in the `CSR` type is now done by simple typedef in the files `qutip/core/data/base.{pyx,pxd}`, so changes it in those two files at compile-time will allow you to switch between 32-bit and 64-bit.  The C++ backing source file `zspmv.cpp` and its associated header are renamed to `matmul_csr_vector` to match the new data-layer naming conventions, but more importantly, the functions are now templated over integral types so they will "magically" work at either size.  In general, `setuptools` likes to link dynamically (on Linux and Mac, but not Windows), and so the templates have to be made explicit, rather than allowing Cython to generate only the necessary one.

_Note:_ I'll make a PR to push deprecation warnings onto `eseries`, `essolve` and all other removals or renames that don't already have one in `master`, ready for when we drop QuTiP 4.6.0.

-----------

_Original top comment:_

Not a complete switch-over to `CSR` from `fast_csr_matrix`, but this goes a really long way to achieving that.  `Qobj` is done completely, but `QobjEvo` is pretty difficult to disentangle, and that really slowed me down a lot, and I'm not at all sure I've got it right there yet.

OpenMP is ignored for now as low-priority.  We'll need to get it working before a QuTiP 5.0 release, but I'm not working on it at the moment.

`CQobjEvo` has been absolutely stripped bare.  All its variants are gone, since the data layer types will handle this instead.  `QobjEvo` is basically managed by shim at the moment, and needs a huge tidy up to bring it up to speed with the new data layer, and keep it maintainable.

The data layer types have been taught to pickle themselves safely and automatically, and so consequently `QobjEvo` is magically pickle-able (aside from the OpenMP types).

This PR includes all the commits from #1296 as well, but it's still a pretty giant PR as it stands right now.

The tests won't pass yet, because I haven't finished converting everything.

jakelishman commented:
This is the same PR as #1328, but resubmitted so the commit history and files changed are simplified after the merging of #1296.

Ericgig commented on qutip/core/cy/cqobjevo.pyx:
```diff
@@ -32,1189 +34,183 @@
-        cdef complex[::1] data_t = np.empty(self.nnz, dtype=complex)
+    cdef void _factor_dynamic(self, double t, Data state) except *:
+        # TODO: why are `dynamic` arguments here when they're purely the
+        # concern of the solvers?
```
QobjEvo are made to remove code duplication in solver, they are not used much anywhere else and don't have much user facing documentation. So it is/were part of the solver not the core.
By making it an independent, user-facing object in core, the solver specific features will need to be moved in a new object.

For now, solvers will be shaky since QobjEvo no longer fit their specific needs.

jakelishman commented:
2dd4680 and b6dcc28 fix the occasional segfaults I was getting, I believe.  The error in `reshape_csr` was probably the randomly appearing one - I assume that for it to cause a segfault, either the output CSR needed to have been allocated on the edge of allocated memory (and then I probably would have seen it in the stack trace with `faulthandler`), or it needed to be called on a sufficiently small matrix that `mem.PyMem_Malloc` allocated into Python stack memory such that the next byte after `out.row_index` pointed into freed Python memory, causing a checksum error on the next `PyMem_Malloc` that hit it, or `PyMem_Free` on a previously created object.

I was able to track down where these were coming from by judicious use of `gcc -fsanitize`, although had we been using Cython's `typedmemoryview`, setting `boundscheck=True` would also have caught it.  Likewise, the error in `permute.dimensions_csr` would have been caught by setting `cdivision=False`, but I just didn't think to try that.  It's probably not worth swapping to memory views just because of this, though - I was able to find it easily enough with `AddressSanitizer`, and memory views have a very non-trivial overhead on initialisation for our use case.

jakelishman commented:
@Ericgig b7cd10a creates a new `qutip.solve` package, and I moved all the solvers and most of their utilities into there.  I put Cython utility modules into the same directory, and just prefixed their names with underscores to mark them as internal, rather than making another new `cy` directory.  I also moved the tests of the solvers into their own `tests/solve` directory, so we can consider having `pytest` run the tests in order at some point as well.

Ericgig commented on qutip/tests/core/test_expect.py:
```diff
@@ -154,7 +154,7 @@ def test_equivalent_to_matrix_element(hermitian):
     op = qutip.rand_herm(dimension, 0.2)
     if not hermitian:
         op = op + 1j*qutip.rand_herm(dimension, 0.1)
-    expected = (state.dag() * op * state).data[0, 0]
```
Not sure if it was what I had in mind anymore, but I expect your removed the scipy 1.13.0 patch where for that release only csr subscription returned 0d numpy array instead of number.

jakelishman replied:
Yeah, it changes in 45992d5.

jakelishman commented:
Yeah, you're right, I should have done the style changes in a different PR.  It was mostly a lack of thought - I have a linter hooked into my text editor, so I get gutter-marks telling me all the PEP8 inconsistencies and stuff like that, and I usually change them without thinking.  The `assert_` usually gets changed to bare `assert` when I'm actually trying to debug the tests - the bare `assert` gives _far_ more debugging detail with `pytest`, so it's much more helpful.

By the way, if you're interested, the list of files which _didn't_ change is
```text
.codeclimate.yml
.coveragerc
.github/ISSUE_TEMPLATE/bug_report.md
.github/ISSUE_TEMPLATE/feature_request.md
.github/ISSUE_TEMPLATE/others.md
.github/pull_request_template.md
.mailmap
.travis.yml
CODE_OF_CONDUCT.md
LICENSE.txt
README.md
pyproject.toml
qutip.bib
qutip/_mkl/__init__.py
qutip/_mkl/spmv.py
qutip/_mkl/spsolve.py
qutip/_mkl/utilities.py
qutip/about.py
qutip/cite.py
qutip/configspec.ini
qutip/control/__init__.py
qutip/control/cy_grape.pyx
qutip/hardware_info.py
qutip/ipynbtools.py
qutip/logging_utils.py
qutip/matplotlib_utilities.py
qutip/orbital.py
qutip/parallel.py
qutip/qip/__init__.py
qutip/qip/algorithms/__init__.py
qutip/qip/circuit_latex.py
qutip/qip/compiler/__init__.py
qutip/qip/compiler/cavityqedcompiler.py
qutip/qip/compiler/gatecompiler.py
qutip/qip/compiler/spinchaincompiler.py
qutip/qip/device/__init__.py
qutip/qip/gates.py
qutip/qip/operations/__init__.py
qutip/qip/qasm.py
qutip/qip/qip_deprecation.py
qutip/tests/Hadamard_params.ini
qutip/tests/__init__.py
qutip/tests/conftest.py
qutip/tests/pytest.ini
qutip/tests/qasm_files/bracket_error.qasm
qutip/tests/qasm_files/command_error.qasm
qutip/tests/qasm_files/qasm_error.qasm
qutip/tests/qasm_files/teleportation.qasm
qutip/tests/qasm_files/test_add.qasm
qutip/tests/qasm_files/test_custom_gates.qasm
qutip/tests/test_control_pulseoptim.py
qutip/tests/test_entropy.py
qutip/tests/test_fileio.py
qutip/tests/test_parallel.py
qutip/tests/test_partial_transpose.py
qutip/tests/test_qasm.py
qutip/tests/test_qft.py
qutip/tests/test_qpt.py
qutip/tests/test_utilities.py
qutip/ui/__init__.py
qutip/ui/progressbar.py
qutip/utilities.py
```

Ericgig commented:
Wow
Great work, merging.

jakelishman commented:
> [![Coverage Status](https://camo.githubusercontent.com/5b95d685d6392d8f3beccad1e575f46ea5187ccd/68747470733a2f2f636f766572616c6c732e696f2f6275696c64732f33323635363130332f6261646765)](https://coveralls.io/builds/32656103)
> 
> Coverage decreased (-10.2%) to 60.763% when pulling **[3b09002](https://github.com/qutip/qutip/commit/3b090027a58ee842c419244920d65623d70ee4a2) on jakelishman:core-replace-fast_csr_matrix** into **[0740a04](https://github.com/qutip/qutip/commit/0740a04cf0b9286b5e13cb0bcf026de6b284c011) on qutip:dev.major**.

It should be noted that the coverage drop is so large here because large tracts of tested Python source files were removed and replaced by _Cython_ source files, which Coveralls does not handle correctly.  These files typically are tested (albeit not yet completely), but Coveralls reports `.pyx` files as having 7 lines of code, of which all 7 are missed.
